src/ucs/sys/ucs_netlink.c

```diff
@@ -0,0 +1,238 @@
+
+err_close_socket:
+    ucs_close_fd(fd_p);
+err:
+    return status;
+}
+
+ucs_status_t ucs_netlink_send_cmd(int protocol, unsigned short nlmsg_type,
+                                  void *nl_protocol_hdr,
+                                  size_t nl_protocol_hdr_size,
+                                  char *recv_msg_buf, size_t *recv_msg_buf_len)
+{
+    struct nlmsghdr nlh = {0};
+    ucs_status_t status;
+    int fd;
+    struct iovec iov[2];
+    size_t bytes_sent;
+
+    status = ucs_netlink_socket_init(&fd, protocol);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    nlh.nlmsg_len   = NLMSG_LENGTH(nl_protocol_hdr_size);
+    nlh.nlmsg_type  = nlmsg_type;
+    nlh.nlmsg_flags = NLM_F_REQUEST | NLM_F_DUMP;
+    iov[0].iov_base = &nlh;
+    iov[0].iov_len  = sizeof(nlh);
+    iov[1].iov_base = nl_protocol_hdr;
+    iov[1].iov_len  = nl_protocol_hdr_size;
+
+    do {
+        status = ucs_socket_sendv_nb(fd, iov, 2, &bytes_sent);
+    } while (status == UCS_ERR_NO_PROGRESS);
+
+    if (status != UCS_OK) {
+        ucs_error("failed to send netlink message (%s)",
+                  ucs_status_string(status));
+        goto out;
+    }
+
+    do {
+        status = ucs_socket_recv_nb(fd, recv_msg_buf, recv_msg_buf_len);
+    } while (status == UCS_ERR_NO_PROGRESS);
+
+    if (status != UCS_OK) {
+        ucs_error("failed to receive netlink message");
+        goto out;
+    }
+
+out:
+    ucs_close_fd(&fd);
+    return status;
+}
+
+ucs_status_t ucs_netlink_parse_msg(void *msg, size_t msg_len,
+                                   ucs_netlink_parse_cb_t parse_cb, void *arg)
+{
+    ucs_status_t status = UCS_INPROGRESS;
+    struct nlmsghdr *nlh;
+
+    for (nlh = (struct nlmsghdr*)msg;
+         (status == UCS_INPROGRESS) && NLMSG_OK(nlh, msg_len) &&
+         nlh->nlmsg_type != NLMSG_DONE;
+         nlh = NLMSG_NEXT(nlh, msg_len)) {
+        if (nlh->nlmsg_type == NLMSG_ERROR) {
+            struct nlmsgerr *err = (struct nlmsgerr *)NLMSG_DATA(nlh);
+            ucs_error("failed to parse netlink message header (%d)", err->error);
+            return UCS_ERR_IO_ERROR;
+        }
+
+        status = parse_cb(nlh, NLMSG_DATA(nlh), arg);
+    }
+
+    return UCS_OK;
+}
+
+struct ucs_netlink_route_info_t {
+    const struct sockaddr *sa_remote;
+    int                    if_index;
+    int                    matching;
+};
+
+
+static ucs_status_t ucs_rtnetlink_get_route_info(struct rtattr *rta, int len,
+                                                 int *if_idx, void **dst_in_addr)
+{
+    *if_idx      = -1;
+    *dst_in_addr = NULL;
+
+    for (; RTA_OK(rta, len); rta = RTA_NEXT(rta, len)) {
+        if (rta->rta_type == RTA_OIF) {
+            *if_idx = *((int*)RTA_DATA(rta));
+        } else if (rta->rta_type == RTA_DST) {
+            *dst_in_addr = RTA_DATA(rta);
+        }
+    }
+
+    if ((*if_idx == -1) || (*dst_in_addr == NULL)) {
+        ucs_debug("either iface index or dest address are missing "
```

Comment:
`ucs_diag` + add specific info in the string

Reply:
`*dst_in_addr` is `void *` pointing to a "raw" IP address, so in order to print it I need to pass its family type from outside of the function. Is it worth it?